File: app/src/main/java/com/cazimir/relaxoo/ui/create_sound/CreateSoundViewModel.java
package com.cazimir.relaxoo.ui.create_sound;

import android.os.Environment;
import android.util.Log;

import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.cazimir.relaxoo.model.Recording;

import java.io.File;
import java.util.ArrayList;

public class CreateSoundViewModel extends ViewModel {

  private static final String TAG = "CreateSoundViewModel";

  private File[] recordings;
  private MutableLiveData<ArrayList<Recording>> recordingsLive = new MutableLiveData<>();

  public CreateSoundViewModel() {
    listChanged();
  }

  public MutableLiveData<ArrayList<Recording>> getRecordingsLive() {
    return recordingsLive;
  }

  public void listChanged() {
    File ownSoundsFolder = Environment.getExternalStoragePublicDirectory("Relaxoo/own_sounds");

    if (!ownSoundsFolder.exists()) {
      ownSoundsFolder.mkdir();
    }

    recordings = ownSoundsFolder.listFiles();

    ArrayList<Recording> recordingsList = new ArrayList<>();

    for (File file : recordings) {
      recordingsList.add(new Recording.Builder().withFile(file).withFileName(file.getName()).build());
    }

    recordingsLive.setValue(recordingsList);
  }

  void deleteRecording(Recording recording) {
    File ownSoundsFolder = Environment.getExternalStoragePublicDirectory("Relaxoo/own_sounds");
    File file2 = new File(ownSoundsFolder, recording.getFile().getName());
    boolean deleted = recording.getFile().delete();
    Log.d(TAG, "deleteRecording() called with: " + deleted);
    if (deleted) {
      listChanged();
    }

  }

  void editRecording(Recording recording, String newName) {
    File ownSoundsFolder = Environment.getExternalStoragePublicDirectory("Relaxoo/own_sounds");
    File from = new File(ownSoundsFolder, recording.getFile().getName());
    File to = new File(ownSoundsFolder, newName + ".wav");

    boolean renamed = from.renameTo(to);

    if (renamed) {
      listChanged();
    }


  }
}

File: app/src/main/java/com/cazimir/relaxoo/ui/create_sound/CreateSoundViewModel.kt
package com.cazimir.relaxoo.ui.create_sound

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.cazimir.relaxoo.eventbus.EventBusUpdateRecordingName
import com.cazimir.relaxoo.model.Recording
import com.cazimir.relaxoo.repository.IRecordingRepository
import org.greenrobot.eventbus.EventBus
import java.io.File
import java.util.*

class CreateSoundViewModel : ViewModel() {

    companion object {
        private const val TAG = "CreateSoundViewModel"
    }


    private val _recordingsLive = MutableLiveData<ArrayList<Recording>>()

    val recordingsLive: LiveData<ArrayList<Recording>>
        get() = _recordingsLive

    lateinit var repository: IRecordingRepository

    fun refreshList() {

        val recordings: Array<File> = repository.getRecordings() ?: mutableListOf<File>() as Array<File>

        val recordingsList = mutableListOf<Recording>()

        // after changing to viewpager2 the recordings returned null so I needed a safe call here
        recordings.let {
            Arrays.sort(recordings) { f1, f2 -> f1.lastModified().compareTo(f2.lastModified()) }
            recordings.reverse()
            for (file in recordings) {
                recordingsList.add(Recording.Builder().withId(file.name).withFile(file).withFileName(file.name).build())
            }
        }

        _recordingsLive.value = recordingsList as ArrayList<Recording>
    }

    fun deleteRecording(recording: Recording) {
        val deleteRecording = repository.deleteRecording(recording)
        if (deleteRecording) refreshList()
    }

    fun editRecording(recording: Recording, newName: String) {
        val result = repository.editRecording(recording, newName)
        if (result) {
            refreshList()
            EventBus.getDefault().post(EventBusUpdateRecordingName(recording.id, newName))
        }
    }
}
